policy-controller/k8s/status/src/index.rs

```diff
@@ -0,0 +1,246 @@
+            "status": status,
+        });
+
+        // Patch the HTTPRoute with its status.
+        let api = k8s::Api::<k8s::policy::HttpRoute>::namespaced(
+            self.client.clone(),
+            &route_id.namespace,
+        );
+        if let Err(error) = api
+            .patch_status(&route_id.name, &patch_params, &k8s::Patch::Merge(patch))
+            .await
+        {
+            tracing::error!(namespace = %route_id.namespace, name = %route_id.name, %error, "Failed to patch HTTPRoute");
+        }
+    }
+
+    async fn process_server_update(&mut self, patch_params: k8s::PatchParams) {
+        let route_ids: Vec<ResourceId> = self.index.read().http_routes.keys().cloned().collect();
+        for route_id in route_ids {
+            self.process_http_route_update(route_id, patch_params.clone())
+                .await;
+        }
+    }
+}
+
+impl Index {
+    pub fn shared(updates: UnboundedSender<Update>) -> SharedIndex {
+        Arc::new(RwLock::new(Self {
+            updates,
+            http_routes: HashMap::new(),
+            servers: HashSet::new(),
+        }))
+    }
+
+    fn update_http_route(&mut self, route_id: ResourceId, route_binding: RouteBinding) -> bool {
+        match self.http_routes.entry(route_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(route_binding);
+            }
+            Entry::Occupied(mut entry) => {
+                if *entry.get() == route_binding {
+                    return false;
+                }
+                entry.insert(route_binding);
+            }
+        }
+        true
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::HttpRoute> for Index {
+    fn apply(&mut self, resource: k8s::policy::HttpRoute) {
+        let namespace = resource
+            .namespace()
+            .expect("HTTPRoute must have a namespace");
+        let name = resource.name_unchecked();
+        let id = ResourceId::new(namespace.clone(), name.clone());
+
+        let route_binding = match RouteBinding::try_from(resource) {
+            Ok(binding) => binding,
+            Err(error) => {
+                tracing::info!(%namespace, %name, %error, "Ignoring HTTPRoute");
+                return;
+            }
+        };
+
+        if self.update_http_route(id.clone(), route_binding) {
+            if let Err(error) = self.updates.send(Update::HttpRoute(id.clone())) {
+                tracing::error!(%id.namespace, %id.name, %error, "Failed to send HTTPRoute update")
+            }
+        }
+    }
+
+    fn delete(&mut self, namespace: String, name: String) {
+        let id = ResourceId::new(namespace, name);
+        self.http_routes.remove(&id);
+    }
+
+    // Since apply only reindexes a single HTTPRoute at a time, there's no need
+    // to handle resets specially.
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::Server> for Index {
+    fn apply(&mut self, resource: k8s::policy::Server) {
```

Comment:
When we get an update here, we synchronously update the index and then queue an update.  Upon asynchronously processing the update, we read from the index and construct the patch.  This means there's a chronological game between when we update the index and when we read from it, potentially leading to some unexpected behavior if the contents of the index have changed in the meantime.

I think this would be easier to reason about if we do as much work synchronously as we can.  i.e. update the index, determine the desired state, and queue that as an update so that we don't need to interact with the index at all in the async processor.

Reply:
Yea this sounds good to me. Originally, more of the work that happens in `fn process_updates` was async, so generating the patch had to happen there. But now that we can synchronously generate the patch, it does make sense to do here and then just send that patch to be asynchronously applied.